std/experimental/allocator/building_blocks/ascending_page_allocator.d

```diff
@@ -0,0 +1,418 @@
+                munmap(data, numPages * pageSize);
+            }
+            else version(Windows)
+            {
+                import core.sys.windows.windows : VirtualFree, MEM_RELEASE;
+                VirtualFree(data, 0, MEM_RELEASE);
+            }
+            data = null;
+        }
+    }
+
+    /**
+    Returns the available size for further allocations in bytes.
+    */
+    size_t getAvailableSize()
+    {
+        return numPages * pageSize + data - offset;
+    }
+
+    /**
+    If the passed buffer is not the last allocation, then ``delta`` can be
+    at most the number of bytes left on the last page.
+    Otherwise, we can expand the last allocation until the end of the virtual
+    address range.
+    */
+    bool expand(ref void[] b, size_t delta)
+    {
+        import std.exception : enforce;
+
+        if (!delta) return true;
+        if (!b.ptr) return false;
+
+        size_t goodSize = goodAllocSize(b.length);
+        size_t bytesLeftOnPage = goodSize - b.length;
+        if (b.ptr + goodSize != offset && delta > bytesLeftOnPage)
+            return false;
+
+        size_t extraPages = 0;
+
+        if (delta > bytesLeftOnPage)
+        {
+            extraPages = goodAllocSize(delta - bytesLeftOnPage) / pageSize;
+        }
+        else
+        {
+            b = cast(void[]) b.ptr[0 .. b.length + delta];
+            return true;
+        }
+
+        if (extraPages > numPages)
+            return false;
+
+        if (offset - data > pageSize * (numPages - extraPages))
+            return false;
+
+        version(Posix)
+        {
+            import core.sys.posix.sys.mman : mprotect, PROT_READ, PROT_WRITE;
+            auto ret = mprotect(offset, extraPages * pageSize, PROT_READ | PROT_WRITE);
+            enforce(ret == 0, "Failed to expand, mprotect failure");
+        }
+        else version(Windows)
+        {
+            import core.sys.windows.windows : VirtualProtect, PAGE_READWRITE;
+            uint oldProtect;
+            auto ret = VirtualProtect(offset, extraPages * pageSize, PAGE_READWRITE, &oldProtect);
+            enforce(ret != 0, "Failed to expand, VirtualProtect failure");
+        }
+
+        pagesUsed += extraPages;
+        offset += extraPages * pageSize;
+        b = cast(void[]) b.ptr[0 .. b.length + delta];
+        return true;
+    }
+
+    /**
+    First it tries to ``expand`` to satisfy ``newSize``.
+    If not possible, it allocates a new buffer of length ``newSize``,
+    copies the contents and deallocates the previous buffer.
+    */
+    bool reallocate(ref void[] b, size_t newSize)
+    {
+        if (!newSize) return deallocate(b);
+        if (!b) return true;
+
+        if (newSize >= b.length && expand(b, newSize - b.length))
+            return true;
+
+        void[] newB = allocate(newSize);
+        if (newB.length <= b.length) newB[] = b[0 .. newB.length];
+        else newB[0 .. b.length] = b[];
+        deallocate(b);
```

Comment:
Could we somehow make use of holes left by this?

Reply:
Interestingly if we were to e.g. store a bit per page, the memory overhead would be smallish (1byte per 32 kbytes) but we can find adjucent free pages when calling expand easily.

Other strategies are possible as well.

Reply:
We don't want to reuse holes. Recall that one important purpose of this allocator is to help with use-after-free.